internal/pkg/util/machine/machine.go

```diff
@@ -0,0 +1,302 @@
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x16},
+	},
+	{
+		Arch:       "ppc64",
+		Sif:        sif.HdrArchPPC64,
+		Machine:    elf.EM_PPC64,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x15},
+	},
+	{
+		Arch:       "ppc64le",
+		Sif:        sif.HdrArchPPC64le,
+		Machine:    elf.EM_PPC64,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x15, 0x00},
+	},
+	{
+		Arch:       "mips",
+		Sif:        sif.HdrArchMIPS,
+		Compatible: "mips64",
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x01, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08},
+	},
+	{
+		Arch:       "mipsle",
+		Sif:        sif.HdrArchMIPSle,
+		Compatible: "mips64le",
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x01, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08, 0x00},
+	},
+	{
+		Arch:       "mips64",
+		Sif:        sif.HdrArchMIPS64,
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08},
+	},
+	{
+		Arch:       "mips64le",
+		Sif:        sif.HdrArchMIPS64le,
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08, 0x00},
+	},
+}
+
+// ArchFromElf returns the architecture string after inspection of the
+// provided elf binary.
+func ArchFromElf(binary string) (string, error) {
+	e, err := elf.Open(binary)
+	if err != nil {
+		return "", fmt.Errorf("failed to open elf binary %s: %s", binary, err)
+	}
+	defer e.Close()
+
+	for _, f := range formats {
+		if f.Machine == e.Machine && f.Class == e.Class && f.Endianness == e.ByteOrder {
+			return f.Arch, nil
+		}
+	}
+
+	return "", ErrUnknownArch
+}
+
+// ArchFromContainer walks through a container filesystem until it
+// find an elf binary to read target architecture from and returns it.
+// If there is no suitable elf binary or if the architecture is not
+// recognized it will return an empty string.
+func ArchFromContainer(container string) string {
+	arch := ""
+
+	filepath.Walk(container, func(path string, info os.FileInfo, err error) error {
```

Comment:
I'm not sure this is the best way to handle this;
 1. This process seems like it could take a while.
 2. What if the container has one or more binaries that are on a different architecture.

So, what about just looking at `/bin/sh`? Since that is the _only_ binary that we care about, should't we just check that?

Reply:
It's true that `/bin/sh` should be the first to look at but many images especially from ngc/docker registries come without `/bin/sh` inside that's why the image is scanned but looking at `/bin/sh` first would clearly optimized the process.

This process doesn't take so long as it returns immediately when it find a suitable ELF binary and the process occurs at build step.
After thinking if there is a binary for a different architecture this is for a very specific use case which should be very rare and this potential case should be minimized by looking at `/bin/sh` first.

Another solution could be to extend use of `--arch` to local build image instead of `--remote` only but I don't like this idea too much as it's more error prone than the above because users could easily define wrong architecture and/or be annoyed by the "non standardization" of architecture flavors.

Overall this is process we can automate to make user experience better.

Reply:
> come without /bin/sh inside

Wait... wait wait wait.... What?!?

Is it that /bin/sh is supposed to be a symlink and that symlink was never created? (To dash, bash, or whatever...)

Or are they doing this as cut down images, and the binary being executed is the entrypoint ... so they remove everything else? So, you effectively have singularity 1.x images that were ldd walks of a given binary?

Reply:
@jms - very minimal container images are becoming a bit more common - especially with the uptake of Go for projects and being able to have static binaries. Eventually when we have metadata outside of the rootfs we can hopefully support these fully.

Reply:
@jmstover Yes micro-service images are often like that, they just embed the binary + libraries + conf/data files and let's go with this binary as entrypoint

Reply:
The current code looking at the shell first seems to be workable here.